airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/common.py

```diff
@@ -0,0 +1,155 @@
+        BasicAmazonAdsStream.__init__(self, config, context=context or SourceContext())
+        HttpStream.__init__(self, *args, **kwargs)
+
+    @property
+    def url_base(self):
+        return self._url
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        return None
+
+    def request_headers(self, *args, **kvargs) -> MutableMapping[str, Any]:
+        return {"Amazon-Advertising-API-ClientId": self._config.client_id}
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        """
+        :return an object representing single record in the response
+        """
+        yield from json.loads(response.text) if response else []
+
+    def _send_request(self, request: requests.PreparedRequest, request_kwargs: Mapping[str, Any]) -> requests.Response:
+        try:
+            return super()._send_request(request, request_kwargs)
+        except requests.exceptions.HTTPError as http_exception:
+            response = http_exception.response
+            if response.status_code in [HTTPStatus.FORBIDDEN]:
+                try:
+                    resp = ErrorResponse.parse_raw(response.text)
+                    self.logger.warn(
+                        f"Unexpected error {resp.code} when processing request {request.url} for {request.headers['Amazon-Advertising-API-Scope']} profile: {resp.details}"
+                    )
+                except ValidationError:
+                    raise http_exception
+            else:
+                raise http_exception
+
+
+class ContextStream(AmazonAdsStream):
+    """
+    Stream for getting resources based on context set by previous stream.
+    """
+
+    flattern_properties = []
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        for record in super().parse_response(response, **kwargs):
+            for prop in self.flattern_properties:
+                if prop in record:
+                    record[prop] = json.dumps(record[prop])
+            yield record
+
+    def read_records(self, *args, **kvargs) -> Iterable[Mapping[str, Any]]:
+        for profile in self._ctx.profiles:
+            self._ctx.current_profile_id = profile.profileId
+            yield from super().read_records(*args, **kvargs)
+
+    def request_headers(self, *args, **kvargs) -> MutableMapping[str, Any]:
+        headers = super().request_headers(*args, **kvargs)
+        headers["Amazon-Advertising-API-Scope"] = str(self._ctx.current_profile_id)
+        return headers
+
+
+class PaginationStream(ContextStream):
+    """
+    Stream for getting resources with pagination support
+    """
+
+    page_size = 100
+
+    def next_page_token(self, response: requests.Response) -> Optional[PageToken]:
+        if not response:
+            return None
+        responses = json.loads(response.text)
+        if len(responses) < self.page_size:
+            self._ctx.current_token = PageToken()
```

Comment:
Why declare a default `PageToken `here, if we will not paginate further in this stream?

Reply:
We need it if we want to run read_records one more time (Actual for profiles stream)

Reply:
oh, this is bad, we don't want to share anything between streams, especially page token, because we might read multiple streams at the same time - see nested streams and slices. pls move current_token if you really need it to stream class